blog/2022/04/simplechains.md

```diff
@@ -0,0 +1,489 @@
+
+But, if you've been a riding the GPU gospel without looking into the details then this plot may be a shocker! However, GPUs are designed as dumb slow chips with many cores, and thus they are only effective on very parallel operations, such as large matrix-matrix multiplications. It is from this point that assumption (2) is derived for large newtork operations. But again, in the case of the small where such GPU kernels will be outperformed by well-designed CPU kernels due to the lack of parallel opportunities
+
+Matrix-matrix operations only occur when batching is able to be used (where each column of the B matrix in A*B is a separate batch). In many cases in scientific machine learning, such as [the calculation of vector-Jacobian products in ODE adjoints](https://youtu.be/6hhF6Llv4sI?t=342), this operation is a matrix-vector multiplication. These operations are smaller and only O(n^2), and as you would guess these effects are amplified in this scenario:
+
+```julia
+using LinearAlgebra, BenchmarkTools, CUDA, LoopVectorization
+
+function mygemmavx!(C, A, B)
+    @tturbo for m ∈ axes(A, 1), n ∈ axes(B, 2)
+        Cmn = zero(eltype(C))
+        for k ∈ axes(A, 2)
+            Cmn += A[m, k] * B[k, n]
+        end
+        C[m, n] = Cmn
+    end
+end
+
+function alloc_timer(n)
+    A = rand(Float32,n,n)
+    B = rand(Float32,n)
+    C = rand(Float32,n)
+    t1 = @belapsed $A * $B
+    t2 = @belapsed (mul!($C,$A,$B))
+    t3 = @belapsed (mygemmavx!($C,$A,$B))
+    A,B,C = (cu(A), cu(B), cu(C))
+    t4 = @belapsed CUDA.@sync($A * $B)
+    t5 = @belapsed CUDA.@sync(mul!($C,$A,$B))
+    t1,t2,t3,t4,t5
+end
+ns = 2 .^ (2:11)
+res = [alloc_timer(n) for n in ns]
+alloc      = [x[1] for x in res]
+noalloc    = [x[2] for x in res]
+noalloclv  = [x[3] for x in res]
+allocgpu   = [x[4] for x in res]
+noallocgpu = [x[5] for x in res]
+
+using Plots
+plot(ns, alloc, label="* (OpenBLAS)", xscale=:log10, yscale=:log10, legend=:bottomright,
+    title="Which Micro-optimizations matter for BLAS2?",
+    yticks=10.0 .^ (-8:0.5:2),
+    ylabel="Time (s)", xlabel="N",)
+plot!(ns,noalloc,label="mul! (OpenBLAS)")
+plot!(ns,noalloclv,label="mygemvavx!")
+plot!(ns,allocgpu,label="* gpu")
+plot!(ns,noallocgpu,label="mul! gpu")
+savefig("microopts_blas2.png")
+```
+
+![](https://user-images.githubusercontent.com/1814174/162625320-310d633a-34bf-407e-8cc9-ec55ca895d83.png)
+
+And remember, the basic operations of a neural network are `sigma.(W*x .+ b)`, and thus there's also an O(n) element-wise operation. As you would guess, this operation becomes more significant as n gets smaller and is requires even more consideration for memory operations. 
+
+```julia
+using LinearAlgebra, BenchmarkTools, CUDA, LoopVectorization
+
+function mybroadcastavx!(C, A, B)
```

Comment:
You could change all the `avx` to `turbo`.

I didn't want to keep answering "Is this x86 only?" questions.
(Although, the SimpleChains.jl example below is about 4x faster on my 4 core laptop than on my M1 mac mini.)